### PR TITLE
Resolve (git) merge conflict: choose which branch to load

### PR DESCRIPTION
**Pull request by amolenaar**
### PR Type
What kind of change does this PR introduce?

- [ ] Bug fix
- [x] Feature
- [ ] Chore (refactoring, formatting, local variables, other cleanup)
- [ ] Documentation content changes

### What is the current behavior?

When a model has been edited on two branches, there's a big change of merge conflicts. It's not possible to resolve those merge conflicts with a text editor (our file format is xml).

One way would maybe be to figure our a merge-friendly file format, but since the model is littered with object references, this seems like a hard job.

Issue Number: #388 

### What is the new behavior?

Gaphor itself should aid in resolving merge conflicts on a semantic level. This should provide the user with the information needed to update a model with incoming changes.

### CODE REVIEW

**Comment by amolenaar** [comment]
This is the solution to merge conflicts at the moment:

image

You can decide to either load the current or the incoming (new) model. 

This means that the change set code (part of this PR) is not actively used. However, we can experiment with it if we want.

File: gaphor/ui/filemanager.py [hunk 1]
```diff
@@ -326,6 +359,31 @@ def response(answer):
             confirm_shutdown()
 
 
+def merge_conflict_dialog(window: Gtk.Window, filename, handler) -> None:
+    dialog = Adw.MessageDialog.new(
+        window,
+        gettext("Resolve merge conflict in model “{filename}”.").format(
```
Comment (danyeaw) [inline gaphor/ui/filemanager.py:365]: Maybe it should say Resolve Merge Conflict?

File: gaphor/ui/filemanager.py [hunk 2]
```diff
@@ -326,6 +359,31 @@ def response(answer):
...
+            filename=filename.name
+        ),
+    )
+    dialog.set_body(
+        gettext(
+            "It looks like the model contains a merge conflict. Do you want to resolve it?"
```
Comment (danyeaw) [inline gaphor/ui/filemanager.py:371]: The model you are opening contains a merge conflict. Do you want to open the current change or the incoming change to the model?

File: gaphor/ui/filemanager.py [hunk 3]
```diff
@@ -326,6 +359,31 @@ def response(answer):
...
+        )
+    )
+    dialog.add_response("cancel", gettext("Cancel"))
+    dialog.add_response("current", gettext("Use current model"))
+    dialog.add_response("incoming", gettext("Use incoming model"))
```
Comment (danyeaw) [inline gaphor/ui/filemanager.py:376]: I think for these two just Current / Incoming, or Current Change / Incoming Change

**Review by danyeaw — APPROVED** [review]
The code looks really good, you made some nice simplifications to how we load models.

I was looking at the [HIG Patterns for Dialogs](https://developer.gnome.org/hig/patterns/feedback/dialogs.html), are we sure that showing a confirmation dialog is the best for this? Maybe we can fix this for the next iteration with manual being able to merge the changes.

Also, when I click Cancel, Gaphor is crashing with `terminated by signal SIGSEGV (Address boundary error)`

File: gaphor/ui/filemanager.py (hunk 2 above)
Comment (amolenaar) [reply gaphor/ui/filemanager.py:371]: That's better indeed.

File: gaphor/ui/filemanager.py (hunk 3 above)
Comment (amolenaar) [reply gaphor/ui/filemanager.py:376]: Shouldn't it contain one verb at least? It's an action after all?

File: gaphor/ui/statuswindow.py [hunk 4]
```diff
@@ -76,13 +64,8 @@ def create_gtk3_window_with_progress_bar(self, frame, vbox, label):
         vbox.pack_start(label, True, True, 0)
         vbox.pack_start(self.progress_bar, expand=False, fill=False, padding=0)
 
+    @g_async()
```
[the remark this answers is not in the document]
Comment (amolenaar) [reply gaphor/ui/statuswindow.py:67]: Thanks :)

**Comment by amolenaar** [comment]
> The code looks really good, you made some nice simplifications to how we load models.
> 
> I was looking at the [HIG Patterns for Dialogs](https://developer.gnome.org/hig/patterns/feedback/dialogs.html), are we sure that showing a confirmation dialog is the best for this? Maybe we can fix this for the next iteration with manual being able to merge the changes.

In the end I want to load the model and show an extra component (e.g. as part of the property editor compartment) with merge conflicts. 
The user should be able to choose which changes to incorporate. I've layed some foundational work in #1961. It still needs a fair amount of work though.

Is that what you're referring to?

> Also, when I click Cancel, Gaphor is crashing with `terminated by signal SIGSEGV (Address boundary error)`

That may be related to #1962. I pushed a fix.

File: gaphor/ui/filemanager.py (hunk 3 above)
Comment (danyeaw) [reply gaphor/ui/filemanager.py]: That is good, I think it would look better if each word was capitalized though

Comment (amolenaar) [reply gaphor/ui/filemanager.py]: Indeed, it should, according to the HIG.

File: gaphor/ui/filemanager.py (hunk 1 above)
Comment (amolenaar) [reply gaphor/ui/filemanager.py:365]: I changed the name to `resolve_merge_conflict_dialog`.

Comment (danyeaw) [reply gaphor/ui/filemanager.py:365]: This is what I meant, the actual heading of the message dialog. 
image

Comment (amolenaar) [reply gaphor/ui/filemanager.py:365]: Ah. That's a lot nicer!

Comment (amolenaar) [reply gaphor/ui/filemanager.py]: fixed

**Pull request merged**